onnxscript/rewriter/ort_fusions/rotary_embedding.py

```diff
@@ -66,22 +71,23 @@ def pattern(self, op, x, end1, start2):
         )
         return op.Concat(x_part_1_rope, x_part_2, axis=-1)
 
-    def check(self, op, x, end1, start2, x_part_1_rope, **_):
+    def check(self, op, x, end1, start2, x_part_1_rope, **_) -> pattern.MatchResult:  # type: ignore[name-defined]
+        check_result = pattern.MatchResult()
         end1_value = _ir_utils.get_singleton_value(end1)
         start2_value = _ir_utils.get_singleton_value(start2)
         if not isinstance(end1_value, int) or not isinstance(start2_value, int):
-            return False
+            return check_result.fail("end1 and start2 are not integers.")
```

Comment:
Suggestion: Extend the MatchResult api (eg., the `fail` method or a new one) to optionally specify the failure-causing node. We could potentially also allow failure-causing values ... or even allow a list of node-or-values. Could be helpful.

The message above can be made a bit more explicit ... may be "The end1 value of first slice and start2 of second slice".

Reply:
The MatchResult API currently supports passing a `failure_node` to `fail`, so I just extended it to support a node, value or a collection of nodes and values. I added example usage in sdpa.py. If this looks good, I'll extend it other such `fail` calls stemming from check functions.